src/mlir/Query.h

```diff
@@ -0,0 +1,556 @@
+{
+  /// Index provides a comparator that orders tuples in lexicographical order,
+  /// but following a given field sequence. For example, when applied to pairs,
+  /// `Index<_, 1, 0>` will first compare the second elements, and only if they
+  /// match will it compare the first ones. This is used to provide custom
+  /// orderings for tuples stored in Relations.
+  ///
+  /// The first type parameter, `Compare`, is used as the underlying comparator
+  /// for elements of the tuples. The same comparator type must be compatible
+  /// with all constituent types.
+  ///
+  /// Index supports both `std::tuple` and any class which defines a
+  /// `data() const` which returns a tuple. This allows wrappers around tuples
+  /// to be used with Relation.
+  template<typename Compare, size_t... Keys>
+  struct Index;
+
+  /// This is the base case, for when no key is provided. Since there is nothing
+  /// to compare, all values are considered equal, ie. (left < right) == false.
+  template<typename Compare>
+  struct Index<Compare>
+  {
+    template<typename T>
+    bool operator()(const T& left, const T& right) const
+    {
+      return false;
+    }
+  };
+
+  template<typename Compare, size_t Head, size_t... Tail>
+  struct Index<Compare, Head, Tail...>
+  {
+    template<typename... Ts>
+    bool
+    operator()(const std::tuple<Ts...>& left, const std::tuple<Ts...>& right)
+    {
+      const auto& left_value = std::get<Head>(left);
+      const auto& right_value = std::get<Head>(right);
+      if (Compare()(left_value, right_value))
+        return true;
+      else if (Compare()(right_value, left_value))
+        return false;
+      else
+        return Index<Compare, Tail...>()(left, right);
+    }
+
+    /// Retrieve the "primary" element of the tuple, that is the one that is
+    /// used first during comparisons.
+    template<typename... Ts>
+    static const auto& get_primary(const std::tuple<Ts...>& value)
+    {
+      return std::get<Head>(value);
+    }
+
+    template<typename T>
+    bool operator()(const T& left, const T& right)
+    {
+      return (*this)(left.data(), right.data());
+    }
+
+    template<typename T>
+    static const auto& get_primary(const T& value)
+    {
+      return get_primary(value.data());
+    }
+
+    /// Underlying comparator used on the contents of the tuple.
+    using value_compare = Compare;
+  };
+
+  namespace detail
+  {
+    /// Type trait used to identify instantiations of `Index`.
+    template<typename T>
+    struct is_index : std::false_type
+    {};
+
+    template<typename Compare, size_t... Keys>
+    struct is_index<Index<Compare, Keys...>> : std::true_type
+    {};
+
+    template<typename T>
+    static constexpr bool is_index_v = is_index<T>::value;
+  }
+
+  struct QueryEngine;
+
+  /// Type-erased version of a Relation, used to be stored in a QueryEngine.
+  struct AbstractRelation
+  {
+    AbstractRelation(QueryEngine& engine);
+    virtual bool iterate() = 0;
```

Comment:
Avoid `virtual` classes as they require RTTI and we'll disable RTTI+EH as soon as we don't depend on it in peglib.

Reply:
Does `virtual` really require RTTI? I thought that was only for `dynamic_cast<>`. MLIR is full of virtual methods.

Reply:
Interesting. My impression from many years ago is that compilers would indeed generate RTTI tables from virtual methods. Either this was a part of the standard that was never defined and compilers got smarter, or my memory is just misplaced.

Reply:
`virtual` doesn't require RTTI.  Both GCC and clang will still generate the two RTTI-related vtable entries (for the offset to the superclass vtable and the `type_info` structure) but will set them both to `nullptr` when RTTI is disabled.